hep/proposals/0001-decorators.md

```diff
@@ -0,0 +1,313 @@
+awt = WorkflowTemplate(name="another-workflow-template")
+
+# We assume we can generate Python code for this (Cluster) Workflow Template from e.g `hera generate stubs external-workflow-template`.
+# Ultimately, awt and ewt will behave the same way when their templates are invoked in dag/steps functions.
+ewt = ClusterWorkflowTemplate(name="external-workflow-template")
+
+@awt.script  # Adds a new script template to a workflow template called "another-workflow-template"
+def setup() -> hio.Output:
+    return hio.Output(result="Setting things up")
+
+class ConcatInput(hio.Input):
+    word_a: str
+    word_b: str
+
+class ConcatOutput(hio.Output):
+    value: str
+
+# We assume we can autogenerate stubs such as these for templates in "external-workflow-template"
+@ewt.script
+def concat(concat_input: ConcatInput) -> ConcatOutput:
+    ...
+    # Note: we can optionally return hio.output(concat_input)
+    # again here to run "local" mock versions of this template ref
+    # this will be useful in local testing
+
+# In the case of kebab-case template names or other details, we can pass the extra info to the decorator:
+@ewt.script(name="my-concat-function")
+def my_concat_function(concat_input: ConcatInput) -> ConcatOutput:
+    ...
+
+
+class WorkerInput(hio.Input):
+    value_a: str
+    value_b: str
+
+class WorkerOutput(hio.Output):
+    value: str
+
+
+@wt.entrypoint
+@wt.dag
+def worker(worker_input: WorkerInput) -> WorkerOutput:
+    # Here, the call to `setup` will create a template ref task since it belongs
+    # in a different workflow template even though it is defined inside this file
+    setup_task = setup()
+
+    # concat serves as a stub, and will also become a template ref task
+    task_a = concat(ConcatInput(word_a=worker_input.value_a, word_b=setup_task.result))
+    task_b = concat(ConcatInput(word_a=worker_input.value_b, word_b=setup_task.result))
+    final_task = concat(ConcatInput(word_a=task_a.value, word_b=task_b.value))
+    return WorkerOutput(value=final_task.value)
+```
+
+
+
+## How to teach (OPTIONAL)
+
+If applicable, describe the differences between teaching this to existing users and new users.
+
+
+# Implementation (OPTIONAL)
+
+TODO: Add links about https://github.com/alexmojaki/sorcery#select_from, https://github.com/alexmojaki/executing#libraries-that-use-this and https://github.com/pwwang/python-varname
+
+## Link to the Implementation PR
+
+# Migration (OPTIONAL)
+
+This section should document breaks to public API and breaks in compatibility due to this HEP's proposed changes. In addition, it should document the proposed steps that one would need to take to work through these changes.
+
+# Drawbacks
+
+Why should we **not** do this?
+
+# Alternatives
+
+- What other designs have been considered?
+- Why is this proposal the best?
+- What is the impact of not doing this?
+
+# Prior Art
+
+Discuss prior art, both the good and bad.
```

Comment:
There might be specific elements of the Prefect syntax this is similar to, particularly the DAG worker examples. Might be good to share them here and explain how this improves that experience